VisualPinball.Unity/VisualPinball.Unity/VPT/Bumper/BumperCollider.cs

```diff
@@ -17,28 +17,27 @@
 using Unity.Collections;
 using Unity.Mathematics;
 using VisualPinball.Engine.Game;
+using VisualPinball.Engine.Common;
 
 namespace VisualPinball.Unity
 {
 	internal static class BumperCollider
 	{
 		public static void Collide(ref BallState ball, ref NativeQueue<EventData>.ParallelWriter events,
 			ref CollisionEventData collEvent, ref BumperRingAnimationState ringState, ref BumperSkirtAnimationState skirtState,
-			in ColliderHeader collHeader, in BumperStaticState state, ref Random random)
+			in ColliderHeader collHeader, in BumperStaticState state, ref Random random, ref InsideOfs insideOfs)
 		{
-			var dot = math.dot(collEvent.HitNormal, ball.Velocity); // needs to be computed before Collide3DWall()!
-			var material = collHeader.Material;
-			BallCollider.Collide3DWall(ref ball, in material, in collEvent, in collEvent.HitNormal, ref random); // reflect ball from wall
-
-			if (state.HitEvent && dot <= -state.Threshold) { // if velocity greater than threshold level
-
-				ball.Velocity += collEvent.HitNormal * state.Force; // add a chunk of velocity to drive ball away
-
-				ringState.IsHit = true;
-				skirtState.HitEvent = true;
-				skirtState.BallPosition = ball.Position;
-
-				events.Enqueue(new EventData(EventId.HitEventsHit, collHeader.ItemId, ball.Id, true));
+			var wasBallInside = insideOfs.IsInsideOf(collHeader.ItemId, ball.Id);
+			var isBallInside = !collEvent.HitFlag;
+			if (isBallInside != wasBallInside) {
+				ball.Position += ball.Velocity * PhysicsConstants.StaticTime;
+				if (isBallInside) {
+					insideOfs.SetInsideOf(collHeader.ItemId, ball.Id);
+					events.Enqueue(new EventData(EventId.HitEventsHit, collHeader.ItemId, ball.Id, true));
+				} else {
+					insideOfs.SetOutsideOf(collHeader.ItemId, ball.Id);
+					events.Enqueue(new EventData(EventId.HitEventsUnhit, collHeader.ItemId, ball.Id, true));
+				}

```

Comment:
Can't you do the stuff you're doing in `BumperApi` in here?

Reply:
I could, but then we would be back to hard-coded bumpers that always push the ball away regardless of what the game logic engine has to say about it.

Reply:
Hang on, that's the goal of the bumpers, no? That's even the case with real-world bumpers if I'm not mistaken. What's a use case where you wouldn't want to do that?

Reply:
Yes, you do almost always want the bumper switch to trigger the bumper coil. But I understand the fundamental design of VPE like this: A table is just a bunch of mindless switches and coils, and the game logic engine decides what coils should fire based on the state of the switches. The changes I made align the bumpers with that design. I guess it's a bit pedantic and rarely comes into play. But for example, MPF turns off the bumpers in attract mode and triggers the bumper coil regardless of the switch state in ball search mode.

Reply:
Okay, gotcha. I'll need some more time to go through this. I've pulled your latest two PRs and with ring speed 1 the animation looks like this on my branch (but I can't easily test how it was before):

https://github.com/user-attachments/assets/dba22f41-47ce-4cb0-8c6d-356c75c9c4e6

So here the ball should approach the center of the bumper much further than right now.

Reply:
I agree it doesn't look quite right. The ring speed is probably not the same as it was before, so that might need some tuning to get right. I haven't changed the size of the collider though. So the issue of how close the ball gets to the bumper is not really related to this PR. Regardless, you can change the radius of both colliders in `BumperApi.CreateColliders`. But there are a couple of additional issues here:
- Proportions: The bumper in the video is way too tall compared to the size of the ball. Look at this video: https://www.youtube.com/watch?v=-9JFBaDKxSc The bumper ring is barely above the ball in the starting position. In your video, it's two ball heights above the ground.
- Logic: The ball is redirected when it enters an invisible trigger collider. If you want it to look totally realistic in slow-motion regardless of parameters like bumper height and ball size, you'd probably have to add another collider to the ring and move that down together with the ring, then apply an impulse to the ball the moment it intersects. But this is not worth the effort in my opinion since the difference wouldn't be noticeable at full game speed or from the typical top-down perspective.